conan/tools/gnu/autotools.py

```diff
@@ -0,0 +1,239 @@
+import copy
+import os
+import platform
+
+
+from conans.client import tools
+from conans.client.tools.oss import OSInfo,  cross_building, \
+    detected_architecture, detected_os, get_gnu_triplet, get_target_os_arch, get_build_os_arch
+from conans.client.tools.win import unix_path
+from conans.errors import ConanException
+from conans.model.build_info import DEFAULT_BIN, DEFAULT_INCLUDE, DEFAULT_LIB, DEFAULT_SHARE
+from conans.util.files import get_abs_path
+
+
+class Autotools(object):
+
+    def __init__(self, conanfile, win_bash=False, include_rpath_flags=False):
+        """
+        FIXME: include_rpath_flags CONAN 2.0 to default True? Could break many packages in center
+        """
+        self._conanfile = conanfile
+        self._win_bash = win_bash
+        self._include_rpath_flags = include_rpath_flags
+        self.subsystem = OSInfo().detect_windows_subsystem() if self._win_bash else None
+        self._os = conanfile.settings.get_safe("os")
+        self._os_version = conanfile.settings.get_safe("os.version")
+        self._os_sdk = conanfile.settings.get_safe("os.sdk")
+        self._os_subsystem = conanfile.settings.get_safe("os.subsystem")
+        self._arch = conanfile.settings.get_safe("arch")
+        self._os_target, self._arch_target = get_target_os_arch(conanfile)
+        self._build_type = conanfile.settings.get_safe("build_type")
+        self._compiler = conanfile.settings.get_safe("compiler")
+        self._compiler_version = conanfile.settings.get_safe("compiler.version")
+        self._compiler_runtime = conanfile.settings.get_safe("compiler.runtime")
+
+        # Precalculate build, host, target triplets
+        self.build, self.host, self.target = self._get_host_build_target_flags()
+
+    def _get_host_build_target_flags(self):
+        """Based on google search for build/host triplets, it could need a lot
+        and complex verification"""
+
+        if self._os_target and self._arch_target:
+            try:
+                target = get_gnu_triplet(self._os_target, self._arch_target, self._compiler)
+            except ConanException as exc:
+                self._conanfile.output.warn(str(exc))
+                target = None
+        else:
+            target = None
+
+        if hasattr(self._conanfile, 'settings_build'):
+            os_build, arch_build = get_build_os_arch(self._conanfile)
+        else:
+            # FIXME: Why not use 'os_build' and 'arch_build' from conanfile.settings?
+            os_build = detected_os() or platform.system()
```

Comment:
I'd say, that may be going forward, avoid auto-detecting altogether and specify everything explicitly in profile/settings, to ensure it's reproducible?

Reply:
100% agree, there will be no auto-detection here, only deterministic from profiles.